### DISCUSSION TITLE
Log4Net Configuration Schema error for 1.2.9 version while running on windows services

### DISCUSSION THREAD

**Discussion by mahadevblore** [apache/logging-log4net#219] [General]
I have a windows service which is written in C# and uses the log4net 1.2.9 for logging.

While trying to start the application from services, we get the following error:

Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.Configuration.ConfigurationErrorsException
   at System.Configuration.ConfigurationSchemaErrors.ThrowIfErrors(Boolean)
   at System.Configuration.BaseConfigurationRecord.ThrowIfParseErrors(System.Configuration.ConfigurationSchemaErrors)
   at System.Configuration.BaseConfigurationRecord.ThrowIfInitErrors()
   at System.Configuration.ClientConfigurationSystem.EnsureInit(System.String)

Exception Info: System.Configuration.ConfigurationErrorsException
   at System.Configuration.ConfigurationManager.PrepareConfigSystem()
   at System.Configuration.ConfigurationManager.GetSection(System.String)
   at System.Configuration.PrivilegedConfigurationManager.GetSection(System.String)
   at System.Diagnostics.DiagnosticsConfiguration.GetConfigSection()
   at System.Diagnostics.DiagnosticsConfiguration.Initialize()
   at System.Diagnostics.DiagnosticsConfiguration.get_IndentSize()
   at System.Diagnostics.TraceInternal.InitializeSettings()
   at System.Diagnostics.TraceInternal.WriteLine(System.String)
   at log4net.Util.LogLog.EmitErrorLine(System.String)
   at log4net.Util.LogLog.Error(System.String, System.Exception)
   at log4net.Util.LogLog..cctor()

Exception Info: System.TypeInitializationException
   at log4net.Util.LogLog.Debug(System.String)
   at log4net.Core.LoggerManager..cctor()

Exception Info: System.TypeInitializationException
   at log4net.Core.LoggerManager.GetRepository(System.Reflection.Assembly)
   at log4net.Config.XmlConfigurator.Configure()
   at CustomService.Main(System.String[])

The configuration files has following things for log4net configuration:
```xml

<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <configSections>
    <sectionGroup name="applicationSettings" type="System.Configuration.ApplicationSettingsGroup, System, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089">
    </sectionGroup>
    <sectionGroup name="EncryptionSettings">
      <section name="EncryptionConfig" type="System.Configuration.NameValueSectionHandler,system,Version=1.0.3300.0,Culture=neutral, PublicKeyToken=b77a5c561934e089,Custom=null"/>
      <section name="PGPConfig" type="System.Configuration.NameValueSectionHandler,system,Version=1.0.3300.0,Culture=neutral, PublicKeyToken=b77a5c561934e089,Custom=null"/>
      <section name="PGPErrorCodes" type="System.Configuration.NameValueSectionHandler,system,Version=1.0.3300.0,Culture=neutral, PublicKeyToken=b77a5c561934e089,Custom=null"/>
    </sectionGroup>
    <section name="log4net" type="log4net.Config.Log4NetConfigurationSectionHandler,log4net" />
  </configSections>
  <appSettings>
</appSettings>
<log4net>
    <root>
      <level value="ALL" />
      <appender-ref ref="RollingLogFileAppender" />
      <appender-ref ref="LogFileAppender" />
    </root>
    <appender name="RollingLogFileAppender" type="log4net.Appender.RollingFileAppender">
      <level value="ALL" />
      <file value="logs\\file.log" />
      <appendToFile value="true" />
      <maxSizeRollBackups value="4" />
      <maximumFileSize value="10MB" />
      <rollingStyle value="Size" />
      <staticLogFileName value="true" />
      <layout type="log4net.Layout.PatternLayout">
        <header value="[Header]&#xD;&#xA;" />
        <footer value="[Footer]&#xD;&#xA;" />
        <conversionPattern value="%date [%thread] %-5level %logger [%ndc] - %message%newline" />
      </layout>
    </appender>
    <appender name="LogFileAppender" type="log4net.Appender.FileAppender">
      <file value="logs\\file.xml" />
      <appendToFile value="true" />
      <layout type="log4net.Layout.PatternLayout">
        <header value="[Header]&#xA;" />
        <footer value="[Footer]&#xA;" />
        <conversionPattern value="%date [%thread] %-5level %logger [%ndc] &lt;%property{auth}&gt; - %message%newline" />
      </layout>
      <layout type="log4net.Layout.XMLLayout" />
    </appender>
  </log4net>
</configuration>
```

The exe name and exe.config files names are the same
customService.exe and customService.exe.config

Any idea why this error is coming up?

I tried multiple things like changing to different config settings by looking at documentation example but none of them checks out.

Please help with this.

**NOTE: The code repository is inaccessible so any code changes are not possible at this point of time.**

**Comment by FreeAndNil** [comment]
@mahadevblore can you post your config file?
BTW log4net 1.x is out of support for 12 years now - that's a very long time.

**Reply by mahadevblore** [reply]
Thanks @FreeAndNil for the reply.

But right now, we are just trying to figure out the underlying issue.

Any help in knowing the root cause for this error in that stack trace would be helpful for us.

Please guide us on how to proceed.

**Reply by FreeAndNil** [reply]
> @mahadevblore can you post your config file?

@mahadevblore see my previous comment

**Comment by mahadevblore** [comment]
Sorry, I thought I had put the config file contents in the discussion thread. 

Here we go:
```xml

<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <configSections>
    <sectionGroup name="applicationSettings" type="System.Configuration.ApplicationSettingsGroup, System, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089">
    </sectionGroup>
    <sectionGroup name="EncryptionSettings">
      <section name="EncryptionConfig" type="System.Configuration.NameValueSectionHandler,system,Version=1.0.3300.0,Culture=neutral, PublicKeyToken=b77a5c561934e089,Custom=null"/>
      <section name="PGPConfig" type="System.Configuration.NameValueSectionHandler,system,Version=1.0.3300.0,Culture=neutral, PublicKeyToken=b77a5c561934e089,Custom=null"/>
      <section name="PGPErrorCodes" type="System.Configuration.NameValueSectionHandler,system,Version=1.0.3300.0,Culture=neutral, PublicKeyToken=b77a5c561934e089,Custom=null"/>
    </sectionGroup>
    <section name="log4net" type="log4net.Config.Log4NetConfigurationSectionHandler,log4net" />
  </configSections>
  <appSettings>
   
  </appSettings>
  <log4net>
   
    <root>
      <level value="ALL" />
      <appender-ref ref="RollingLogFileAppender" />
      <appender-ref ref="LogFileAppender" />
    </root>
    <appender name="RollingLogFileAppender" type="log4net.Appender.RollingFileAppender">
      <level value="ALL" />
      <file value="logs\\OutgoingFileMgmt.log" />
      <appendToFile value="true" />
      <maxSizeRollBackups value="4" />
      <maximumFileSize value="10MB" />
      <rollingStyle value="Size" />
      <staticLogFileName value="true" />
      <layout type="log4net.Layout.PatternLayout">
        <header value="[Header]&#xD;&#xA;" />
        <footer value="[Footer]&#xD;&#xA;" />
        <conversionPattern value="%date [%thread] %-5level %logger [%ndc] - %message%newline" />
      </layout>
    </appender>
    <appender name="LogFileAppender" type="log4net.Appender.FileAppender">
      <file value="logs\\OutgoingFileMgmt.xml" />
      <appendToFile value="true" />
      <layout type="log4net.Layout.PatternLayout">
        <header value="[Header]&#xA;" />
        <footer value="[Footer]&#xA;" />
        <conversionPattern value="%date [%thread] %-5level %logger [%ndc] &lt;%property{auth}&gt; - %message%newline" />
      </layout>
      <layout type="log4net.Layout.XMLLayout" />
    </appender>
  </log4net>
  
  <startup>
    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.0" />
  </startup>
  <system.web>
    <membership defaultProvider="ClientAuthenticationMembershipProvider">
      <providers>
        <add name="ClientAuthenticationMembershipProvider" type="System.Web.ClientServices.Providers.ClientFormsAuthenticationMembershipProvider, System.Web.Extensions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" serviceUri="" />
      </providers>
    </membership>
    <roleManager defaultProvider="ClientRoleProvider" enabled="true">
      <providers>
        <add name="ClientRoleProvider" type="System.Web.ClientServices.Providers.ClientRoleProvider, System.Web.Extensions, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35" serviceUri="" cacheTimeout="86400" />
      </providers>
    </roleManager>
  </system.web>
  <system.diagnostics>
    <trace autoflush="true" />
    <sources>
      <source name="System.Net.Sockets">
        <listeners>
          <add name="MyTraceFile"/>
        </listeners>
      </source>
      <source name="System.Net">
        <listeners>
          <add name="MyTraceFile"/>
        </listeners>
      </source>
    </sources>
    <sharedListeners>
      <add name="MyTraceFile"
        type="System.Diagnostics.TextWriterTraceListener"
        initializeData="logs\SshNetTrace.log"/>
    </sharedListeners>
    <switches>
      <add name="System.Net" value="Verbose" />
      <add name="System.Net.Sockets" value="Verbose" />
    </switches>
 </system.diagnostics>
</configuration>
```

**Reply by mahadevblore** [reply]
[config.txt](https://github.com/user-attachments/files/18213032/config.txt)

@FreeAndNil : I was unable to put the complete config in the thread so added the file here. Please help!

**Reply by FreeAndNil** [reply]
@mahadevblore thanks. I will take a look.

**Reply by FreeAndNil** [reply]
@mahadevblore Please remove the empty first line in the config file?
Or is this line only in config.txt?

**Reply by FreeAndNil** [reply]
I've adjusted your comment with
"```xml"

xml content

"```"

**Reply by FreeAndNil** [reply]
Did this application ever work for you?
Can you start you app in a debugger?

**Reply by mahadevblore** [reply]
App is working in Prod and it has not been compiled in ages.

Unfortunately, we are unable to get the source code to compile or debug due to client's confidentiality. This is just lift and shift project only.

**Reply by FreeAndNil** [reply]
It's probably time to tell your client what "not supported" means.
I'm sorry.

**Reply by mahadevblore** [reply]
Here is the complete config file.

[config.txt](https://github.com/user-attachments/files/18217947/config.txt)


one last look? @FreeAndNil to see if its formatting or config file issue?

**Reply by FreeAndNil** [reply]
The config file seems legit.
You can try to remove the log4net section and the corresponding entry in configSections.
If the service starts, it is probably a log4net problem.
If it doesn't start, it is something completely different. Maybe some assembly is missing?
You can take a look at the process on the Prod machine (with process explorer) an check which assemblies are loaded and then double check whether they are present on the new machine.

**Reply by mahadevblore** [reply]
Thanks for the input @FreeAndNil . Will give the process explorer a try as it might give some clue.

**Reply by mahadevblore** [reply]
U were right @FreeAndNil  about removing the log4net sections from config file but the error still persists.

Seems like they are using XMLConfigurator.Config() which is looking for something in config?

Not sure why the error is still in place after removing the config sections and log4net sections from xml.

**Reply by mahadevblore** [reply]
Will try to use process manager from Prod environment to see if we can figure out something.
